lib/delegate.js

```diff
@@ -129,26 +129,63 @@ const delegateToComponent = async function (component, options) {
 
   debug(`delegating ${print(document)} to ${component.name}`);
 
-  let { data, errors = [] } = await execute({
-    document, 
-    schema: component.schema, 
-    rootValue: info.rootValue, 
-    contextValue, 
+  if (info.operation.operation === 'query' || info.operation.operation === 'mutation') {
+    let { data, errors = []} = await execute({
+      document, 
+      schema: component.schema, 
+      rootValue: info.rootValue, 
+      contextValue, 
+      variableValues: info.variableValues
+    });
+
+    if (!data) {
+      data = {};
+    }
+    
+    if (errors.length > 0) {
+      for (const error of errors) {
+        deepSet(data, error.path, error);
+      }
+    }
+
+    return data[targetRootField];
+  }
+
+  const result = await subscribe({
+    document,
+    schema: component.schema,
+    rootValue: info.rootValue,
+    contextValue,
     variableValues: info.variableValues
   });
 
-  // data can be completely null and destructuring defaults only apply to undefined
-  if (!data) {
-    data = {};
-  }
-  
-  if (errors.length > 0) {
-    for (const error of errors) {
-      deepSet(data, error.path, error);
+  if (Symbol.asyncIterator in result) {
+    return {
+      async next() {
+        const nextResult = await result.next();
+        if (nextResult.done) {
+          return nextResult;
+        }
+
+        let { value: { data, errors = []}} = nextResult;
+
+        if (!data) {
+          data = {};
+        }
+
+        if (errors.length > 0) {
+          for (const error of errors) {
+            deepSet(data, error.path, error);
+          }
+        }
+
+        return { done: false, value: { [targetRootField]: nextResult.value.data[targetRootField]}};
+      },
+      [Symbol.asyncIterator]() {
+        return this;
+      }
```

Comment:
Just wanted to understand this part better, are we iterating through `asyncIterator` and returning async objects using the results (with `next` method along with `value` and `done` properties) again in `asyncIterator` to merge errors, in case there is any?

Reply:
close - effectively we check if the result is an async iterator and wrap that async iterator in another async iterator, so that we can do what you said - return the result for the given target field or merge in errors if there are any